### PR TITLE
Improved Repel / Lure Menu

### PR DESCRIPTION
**Pull request by pkmnsnfrn**
# Description
Jaizu wrote a non-public implementation of the Repel / Lure menu. This PR is a refactored version of Jaizu's implementation.

## Definitions
* Spray Type: Pokémon has two types of sprays - a Repel and a Lure. 
* Strength: Each spray type has three different strengths - regular, super, and max.

# Functionality
When a spray's step count reaches 0, the game counts the number of unique sprays of that type in the bag. 

* **`EventScript_SprayWoreOffMessage`**: If the number is zero, the player is in informed their spray has run out.
* **`EventScript_UseDifferentSpray`**: If the number is one, the player is asked if they would to use that spray. 
  * **`EventScript_UseLastUsedSpray`**: If that spray ALSO matches the strength of the previously used spray (requires `VAR_LAST_REPEL_LURE_USED`), the printed message will change.
* **`EventScript_ChooseWhichSpray`**: If the number is greater than two AND `I_REPEL_LURE_MENU` is `TRUE`, a menu opens to allow the player to choose from a list of sprays (or cancel their input). If `VAR_LAST_REPEL_LURE_USED` is defined, the menu's cursor begins at the last used spray's position. The y position of the menu changes depending on the number of sprays in the list.
* If the number is greater than two AND `I_REPEL_LURE_MENU` is `FALSE`, the behavior from `EventScript_UseDifferentSpray` is adopted. The weakest spray is presented as the option.

In any of the aforementioned cases:
* pressing B or Cancel will not use a new spray. 
* Selecting a spray will use that spray and consume the item.
* If `VAR_LAST_REPEL_LURE_USED` is defined, it's name is used when a spray runs out, as opposed to the base name of the spray's type.

# Testing
* [Scripts](https://github.com/PokemonSanFran/pokeemerald-expansion/commit/d53e2717df48d4192e9c65f0d42d61a10fb0c4ee)
* [Save File](https://files.catbox.moe/slmuhd.sav)

# Showcase
## Video
There are videos that answer shows off the behavior under the the following configurations:

* [`VAR_LAST_REPEL_LURE_USED != 0` && `I_REPEL_LURE_MENU == TRUE`](https://www.youtube.com/watch?v=gixQZ_g8qMs)
* [`VAR_LAST_REPEL_LURE_USED != 0` && `I_REPEL_LURE_MENU == FALSE`](https://www.youtube.com/watch?v=g9T9CJuNpS4)
* [`VAR_LAST_REPEL_LURE_USED == 0` && `I_REPEL_LURE_MENU == TRUE`](https://www.youtube.com/watch?v=aE3SuyVIHZc)
* [`VAR_LAST_REPEL_LURE_USED == 0` && `I_REPEL_LURE_MENU == FALSE`](https://www.youtube.com/watch?v=Bf8mVkfN5WQ)

In each video, the player uses the debug scripts to modify their bag and then takes a step to see the resulting behavior. Each debug script in the video corresponds to the following questions:

> What happens when the player's spray has run out and their bag:
> 1. does not a spray of matching type?
> 2. has a spray of matching type and strength?
> 3. has one spray of matching type but not strength?
> 4. has two sprays of matching type
> 5. has three sprays of matching type

# Scope
## Not In Scope
* `ItemUseOutOfBattle_X`
* `Task_StartUseX`
* `Task_UseX`
* [An optimized .inc file](https://www.huderlem.com/poryscript-playground/?switches=LANG=EN%20GAME_VERSION=EMERALD&code=MYewdgzgLgBAagQQEoH0DKAFJCCaN8GGEC88yKADAB4AcFFALAFCiSyKoByAqgLLpZc6ACpIAopwDiwgBIxSHSrXoBWFuGhlUmbDhTCcGMURJaldCgE51bM5zEB1AbpOnF1CwHYmTCMABOAJYADrBiAG4ApmBQaAEhUCgAQgCuAGZpkf5owf4AhgCenHkAtpFMAN5MAMSBaWYAMghowijiRg0oDdziKNxoYgAiMACEpBRMhABG6ZlBUJElYKWRABQtqIoAjAA0ik0tbWIdXT1ifQODAJSTBP6RUCn+YDXRACZ1PoQQwZHAgXkADbhPL+VaKHRCAxGHaSB4NPLQbgQSJvHL5ArCAq-G63fB1GCrEFgiGCPTQsRXeSkACSwjE-HaYgaN0IVVcMBmGSygQWSxW61EKG2OzpDKOHVZBAAvniYJFASi5ezXFy5rzFssyoLNuRdmL+N1xFL8LLCPdHs8mLLfPFQjAItFYnbEujCg4QPcAPIZSpywEgYAAazlwCBgNWjpicSCoWSsyybqKKxNMB+fwBwNB4PIPH4kL0Gwk0hkOxgcKgnBSJSmidyhTQUHuYAA5lAABa474Ad15wHbROzijzziERaksiuqZVJjDKJgFAAXDAWyAoCBI1Foy70PWCh7vRleJEIBA8i3IqnCHPIjAtsvV+vN06Ywk+ii0IFW4DIkmrwQ3kiNI8hSQEoAfNcNyjZ1Y0SGQ8jAN4f14UCoBCH8kwgVNZRtPxYIdLcYLfeDEOQ1D0N-PcID9Wp6hpCVmVOXpeAkbhqRgURuDEOVHygwjXzjABhdsQBAFEHHbQJ+z-V5FXKQheOfbdYPfX8vxbDC9xuap3k+XCXQIl8dyTA9Ih9NJj1Pc9yhnGASggFsphAKhVmESIqFdPdTPMnZeDQSQki9AANdAaUkThU3uH9EUicM5Xea0fDwhJDOUt9kTU79KIxGiCX2ZpWiZTojXOfohnYiZr3DJSiLjDLBjqOYXy02SlWvds-iDDUShzVADkK45GJKi4hh2LZUwJQcSXIcQ0G4BphCpYhSE4yllTlKrAQjaCBMSDKESRFE0RawgzQIBU2oIWzNu2-id3qxqsmajFsNeRDPmvAMUTKKyLzlKLYpROLCAS-T8J24yHk-LKkwQRCMr9b4HmJXqjjmhadgoVMwy2mrdtUmTzQeJ4XjBlKIZUh7uWbWI90Rgh03+IEUeHRxRxwWEob3Gk3lTNUeT5LU1g2YVyAAJj2XM2YLVN7Mc5zXPczyMW8jIEAgIMMqC3z-MCkKcDENBOC9Cb6im1HZvmxb2INtBpw2ghQGCAoWfIAtJa4aWyX-fAcduoyVLQKH1M0jE4beDLXqJy1SZ8WjGgKhjirOEbhjGBdbXBu7KZRA6oAy46ctsuWnJctyPN3FXPTMtWNYyuG1w6-wdYC4KUFt43TcJV3UEthaltIW37Y5J2XaHN2yQ9rpE6KpjSsuH2YD9vHIdiEPssKcPI7lM78AtEnEp0960jjgl6Nn4aWJ4djVsz8ns7fYTRPEyTpLp4uTzPC9FYrkzq-M9WmsUQNw7FkVM3Y8i8h+l-eSjtwzLDQlEVYgx8jdiTMeMAKRwGQKgNAPICwvgEEmj3NGVsqTp14FbeiSQUBYENpcGAAAyRh3dx690NmQ0YChyAeHoAAZinOtEe1UKbpQ-CdGUoYvqLE-tZf6CpAaxS2vFRCh9dInzvvaURdVxFFzlMEQEhQUTrHOEVbBvIlShngfgwISCSJIQ3gUDBKQn5SUvIQveiwQBRG6qjd2d4+YJnmJqAUItFASzMDLOUJcFbl2Vu6f+GQtYtz1qFcKkViZWhtAsDyHElaVwSYeE+tk0iehKPg1YAAiCoYS9TSgAOQQHlNyYAsBuzVxgCADIgBk4AAOMACrKk3BtPHfKhwL4pzKmncYMAAA+syYDx3PoNZOzFWI3yQFxJgOTYBxIKfuRJaRAH1zAI3LI9N8ClP8OUqAVSalCm2A0ppQFMitJgO0+4nSMgjAADpgAyjABCZz-AAH4hlqOPj4HZeTf5eUOcclEQULkwCuTcu5tTUBbCec015bSOldLSH0-pfyAV5BgPc3UqAxbSjBcMqFSsYXxIOUUrWyLUUVOqRgJoBskDShgCkI6MBQF-IpaLKl0oAB04LZRAA)

## In Scope
* All happy path and edge cases are handled / bug free
* When not defined, behaviors involving `I_REPEL_LURE_MENU` and `VAR_LAST_REPEL_LURE_USED` are gracefully handled
* `TryDrawXMenu`, `HandleXMenuChoice`, `EventScript_XWoreOff`, and `EventScript_XUseAnother` are refactored into single functions

# **Discord contact info**
I am `pkmnsnfrn` on Discord, and there is a [discussion thread on Discord](https://discord.com/channels/419213663107416084/1174959238246842418)

# Thanks
* This is heavily based off the work already in expansion, and the new contributions mostly came from @Jaizu

### CODE REVIEW

File: data/specials.inc [hunk 1]
```diff
@@ -536,3 +536,6 @@ gSpecials::
 	def_special TrySetBattleTowerLinkType
 	def_special SetTotemBoost
 	def_special TrySpecialOverworldEvo
+    def_special GetNumberSprayStrength
+    def_special GetSprayId
+    def_special GetLastUsedSprayType
```
Comment (LOuroboros) [inline data/specials.inc:541]: Tabs instead of spaces.

Comment (pkmnsnfrn) [reply data/specials.inc]: https://github.com/rh-hideout/pokeemerald-expansion/pull/3594/commits/8e66726ad87290910c8d8beef5fab22b7cfdf5c2

File: data/scripts/repel.inc [hunk 2]
```diff
@@ -1,117 +1,161 @@
...
-	checkitem ITEM_REPEL, 1
-	goto_if_eq VAR_RESULT, TRUE, EventScript_RepelUseAnother
-	checkitem ITEM_SUPER_REPEL, 1
-	goto_if_eq VAR_RESULT, TRUE, EventScript_RepelUseAnother
-	checkitem ITEM_MAX_REPEL, 1
-	goto_if_eq VAR_RESULT, TRUE, EventScript_RepelUseAnother
-.else
-	checkitem VAR_LAST_REPEL_LURE_USED, 1
-	goto_if_eq VAR_RESULT, TRUE, EventScript_RepelUseAnother
+.set VAR_SPRAY, VAR_0x8004
```
[the remark this answers is not in the document]
Comment (pkmnsnfrn) [reply data/scripts/repel.inc]: 389477217e2c2c8304db10e2d4c6da2936bc6845

File: data/scripts/repel.inc [hunk 3]
```diff
@@ -1,117 +1,161 @@
...
-	msgbox Text_UseAnotherRepel, MSGBOX_YESNO
+	call EventScript_BufferSprayName
+	specialvar VAR_NUM_SPRAY_STRENGTH, GetNumberSprayStrength
+	switch VAR_NUM_SPRAY_STRENGTH
+	case 0, EventScript_SprayWoreOff_2
+	case 1, EventScript_SprayWoreOff_3
+	goto EventScript_HandleMultipleSprays
+	return
+
+EventScript_SprayWoreOff_2:
```
[the remark this answers is not in the document]
Comment (pkmnsnfrn) [reply data/scripts/repel.inc]: 389477217e2c2c8304db10e2d4c6da2936bc6845

File: src/sprays.c [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+#include "script_menu.h"
+#include "menu.h"
+
+#define SPRAY_COUNT 0
+#define SPRAY_GET 1
+
+#define NUM_SPRAY_TYPES 3
+#define SPRAY_MENU_Y_COORD 8
+
+#define VAR_SPRAY gSpecialVar_0x8004
```
[the remark this answers is not in the document]
Comment (pkmnsnfrn) [reply src/sprays.c]: 389477217e2c2c8304db10e2d4c6da2936bc6845

**Review by Bassoonian — APPROVED** [review]

**Pull request merged**